File: log_recorder/src/log_recorder.cpp
#include "log_manager/log_manager.hpp"

LogManager::LogManager(const std::string &nodeName) : Node(nodeName) {
    RCLCPP_INFO(this->get_logger(), "Log Manager Node Initialized");
}

void LogManager::init() {
    isRecording_ = false;
    // Service for starting recording
    startService_ = this->create_service<std_srvs::srv::Trigger>(
        "start_recording",
        std::bind(&LogManager::startRecording, this, std::placeholders::_1, std::placeholders::_2)
    );

    // Service for stopping recording
    stopService_ = this->create_service<std_srvs::srv::Trigger>(
        "stop_recording",
        std::bind(&LogManager::stopRecording, this, std::placeholders::_1, std::placeholders::_2)
    );

    // Subscribe to multiple topics
    tfSub_ = this->create_subscription<tf2_msgs::msg::TFMessage>(
        "tf", 10,std::bind(&LogManager::tfSubCallback, this, std::placeholders::_1));

    tfStaticSub_ = this->create_subscription<tf2_msgs::msg::TFMessage>(
        "tf_static", 10,std::bind(&LogManager::tfStaticSubCallback, this, std::placeholders::_1));


    RCLCPP_INFO(this->get_logger(), "Log Manager Node Initialized");
}

// Service Trigger Functions
void LogManager::startRecording(const std::shared_ptr<std_srvs::srv::Trigger::Request> request, std::shared_ptr<std_srvs::srv::Trigger::Response> response) {
    if (isRecording_) {
        response->success = false;
        response->message = "Recording is already in progress.";
        return;
    }
    // Initialize the recorder
    recorder_ = std::make_unique<rosbag2_cpp::Writer>();

    // Generate unique filename
    std::string filename = generateUniqueFilename();

    // Start recording to filename
    recorder_->open(filename);
    
    // Set recording flag
    isRecording_ = true;

    // Send debug messages
    response->success = true;
    response->message = "Started recording to " + filename;
    RCLCPP_INFO(this->get_logger(), "Started recording to: %s", filename.c_str());
}

void LogManager::stopRecording(const std::shared_ptr<std_srvs::srv::Trigger::Request> request, std::shared_ptr<std_srvs::srv::Trigger::Response> response) {
    if (!isRecording_) {
        response->success = false;
        response->message = "No recording is in progress.";
        return;
    }
    // Stop recording
    recorder_->close();
    // Set recording flag
    isRecording_ = false;

    // Send debug messages
    response->success = true;
    response->message = "Stopped recording.";
    RCLCPP_INFO(this->get_logger(), "Stopped recording.");
}

// Helper Functions
std::string LogManager::generateUniqueFilename() {
    // Declare local vars
    std::stringstream ss;

    // Set up the clock
    auto now = std::chrono::system_clock::now();
    auto in_time_t = std::chrono::system_clock::to_time_t(now);

    // Create a stringstream with the current timestamp
    ss << "log_recording_" << std::put_time(std::localtime(&in_time_t), "%Y%m%d-%H%M%S");// << ".db3";

    // Console Log the stringstream converted to a string
    RCLCPP_INFO(this->get_logger(), "Generated filename: %s", ss.str().c_str());

    return ss.str();
}

// Subscription Callback functions
void LogManager::tfSubCallback(std::shared_ptr<rclcpp::SerializedMessage> msg) {
    rclcpp::Time timestamp = this->now();
    if(isRecording_){
        recorder_->write(msg, "tf", "tf2_msgs/msg/TFMessage", timestamp);
    }
}

void LogManager::tfStaticSubCallback(std::shared_ptr<rclcpp::SerializedMessage> msg) {
    rclcpp::Time timestamp = this->now();
    if(isRecording_){
        recorder_->write(msg, "tf_static", "tf2_msgs/msg/TFMessage", timestamp);
    }
}

File: log_recorder/log_recorder/log_recorder.py
import rclpy
from rclpy.node import Node
from std_srvs.srv import Trigger
from rcl_interfaces.msg import ParameterDescriptor
from rcl_interfaces.msg import ParameterType
from datetime import datetime
import os
import subprocess
import signal


class LogRecorder(Node):
    def __init__(self):
        super().__init__('log_recorder')

        # Declare parameters
        self.declare_parameter('topic_list', [""], 
            ParameterDescriptor(
                type=ParameterType.PARAMETER_STRING_ARRAY,
                description="A list of topics to record",
            )
        )

        # Initialize services
        self.start_service = self.create_service(Trigger, 'start_recording', self.start_recording_callback)
        self.stop_service = self.create_service(Trigger, 'stop_recording', self.stop_recording_callback)

        # Declare members
        self.is_recording = False
        self.process = None

        self.get_logger().info(f"Log Recorder Node is initialized.")

    def start_recording_callback(self, request, response):
        if self.is_recording:
            response.success = False
            response.message = "Recording is already in progress."
            return response

        # Generate unique filename based on date and time
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        filename = f"log_recording_{timestamp}"
        fileDir = "./"
        self.get_logger().info(f"Generated Filename: {fileDir + filename}")

        # Get the topics list from parameters
        topic_list = self.get_parameter('topic_list').get_parameter_value().string_array_value
        if not topic_list:
            response.success = False
            response.message = "No topics specified for recording."
            return response
        
        # Prepare the command to start rosbag recording
        topics = ' '.join(topic_list)  # Join the topics into a single string
        command = f"ros2 bag record {topics} -o {fileDir + filename}"
        
        # The os.setsid() is passed in the argument preexec_fn so it's run after the fork() and before  exec() to run the shell.
        self.process = subprocess.Popen(command, stdout=subprocess.PIPE, shell=True, preexec_fn=os.setsid)

        self.is_recording = True
        response.success = True
        response.message = f"Started recording to: {fileDir + filename}."
        return response

    def stop_recording_callback(self, request, response):
        if not self.is_recording:
            response.success = False
            response.message = "No recording in progress."
            return response

        self.get_logger().info("Preparing to stop....")
        os.killpg(os.getpgid(self.process.pid), signal.SIGINT)
        
        self.is_recording = False

        response.success = True
        response.message = "Stopped recording."

        return response

def main(args=None):
    rclpy.init(args=args)
    node = LogRecorder()

    rclpy.spin(node)

    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
